crates/holochain_zome_types/src/countersigning.rs

```diff
@@ -0,0 +1,288 @@
+    header_seq: u32,
+    /// The signature of all preflight request data and agent state data.
+    signature: Signature,
+}
+
+impl CounterSigningAgentState {
+    /// Agent index accessor.
+    pub fn agent_index(&self) -> u8 {
+        self.agent_index
+    }
+
+    /// Chain top accessor.
+    pub fn chain_top_ref(&self) -> &HeaderHash {
+        &self.chain_top
+    }
+
+    /// Header seq accessor.
+    pub fn header_seq(&self) -> u32 {
+        self.header_seq
+    }
+
+    /// Signature accessor.
+    pub fn signature_ref(&self) -> &Signature {
+        &self.signature
+    }
+}
+
+/// A vector of agents to countersign a shared entry.
+/// The vector must be sorted to generate a CounterSigningTag.
+#[derive(PartialEq, Eq, Debug, Clone, Serialize, Deserialize)]
+pub struct CounterSigningAgentStates(Vec<CounterSigningAgentState>);
+
+/// Enum to mirror Header for all the shared data required to build session headers.
+/// Does NOT hold any agent specific information.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub enum HeaderBase {
+    /// Mirrors Header::Create.
+    Create(CreateBase),
+    /// Mirrors Header::Update.
+    Update(UpdateBase),
+    // @todo - These headers don't have entries so there's nowhere obvious to put the CounterSigningSessionData.
+    // Delete(DeleteBase),
+    // DeleteLink(DeleteLinkBase),
+    // CreateLink(CreateLinkBase),
+}
+
+/// Base data for Create headers.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct CreateBase {
+    entry_type: EntryType,
+    entry_hash: EntryHash,
+}
+
+/// Base data for Update headers.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct UpdateBase {
+    original_header_address: HeaderHash,
+    original_entry_address: EntryHash,
+    entry_type: EntryType,
+    entry_hash: EntryHash,
+}
+
+/// All the data required for a countersigning session.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct CounterSigningSessionData {
+    preflight_request: PreflightRequest,
+    responses: Vec<CounterSigningAgentState>,
+}
+
+/// Build an unsigned Create header from session data, shared create data and an agent's state.
+impl
+    From<(
+        &CounterSigningSessionData,
+        &CreateBase,
+        &CounterSigningAgentState,
+    )> for Create
+{
+    fn from(
+        (session_data, create_base, agent_state): (
+            &CounterSigningSessionData,
+            &CreateBase,
+            &CounterSigningAgentState,
+        ),
+    ) -> Self {
+        Create {
+            author: session_data.preflight_request.signing_agents[agent_state.agent_index as usize]
+                .0
+                .clone(),
+            timestamp: Timestamp(
+                session_data
+                    .preflight_request
+                    .session_times
+                    .start
+                    .0
+                    .checked_add(SESSION_HEADER_TIME_OFFSET_MILLIS)
```

Comment:
This feels weird, declaring that something was created after it actually was, even if by a second. Can we also just put the fudge factor on the other side of the equation, extending the start time back in time by a second when doing the range check?

Reply:
@maackle there's no specific or singular time that the headers "are created", we're mapping from session data to a header set, where the session data is the entry but the headers are something else

the main thing this is guarding against is busy chains failing to accept a preflight due to many writes while the preflight is still propagating over the network